### ISSUE TITLE
Secure Livepeer against single-frame segments attack

### ISSUE THREAD

**Issue by leszko** [livepeer/go-livepeer#2335]
**Is your feature request related to a problem? Please describe.**
We've recently seen a high number of `ZeroSegments` error which was reported in [this GH Issue](https://github.com/livepeer/go-livepeer/issues/2322). While the error itself is related to the [bug in LPMS](https://github.com/livepeer/lpms/issues/311), it also means that users of the Livepeer ecosystem started to send some artificially-created one-frame segments.

These segments could be a result of some testing or malicious behavior (like trying of kind-of DoS attack).

**Describe the solution you'd like**
We need to make Livepeer ready for users sending a lot of one-frame segments, because the payment associated with such segments is very low, but the work required by Os to do is high (session intialization, etc.).

Some ideas as the solution:
- Detect such segments in B and don't send them to Os
- Change the payment details, e.g. not only per pixel, but also pay extra for the first segment/frame

**Comment by iameli** [comment]
This is not unusual in real-world transcoding cases — OBS, upon experiencing network congestion, will discard everything except keyframes. UGC platforms often include lots of users streaming with poor internet, thus we end up with lots of keyframe-only segments. This makes me opposed to

> Detect such segments in B and don't send them to Os

Because these are still very legitimate use cases.

> but the work required by Os to do is high (session intialization, etc.).

True for hardware transcoders; less true for software transcoders. O software could detect single-frame segments and fall back to a software transcoder? It's a single frame, so processing overhead should be relatively minimal, no need to calculate motion vectors and whatnot.

**Comment by MikeIndiaAlpha** [comment]
I hope it isn't bad place for a comment. Are you sure guys that the segments in question contain _any_ frame at all? Because for example if the .h264 stream in question contains just a first slice of multi slice IDR picture, then it is not enough for the decoder to create any output at all. Any chance I could have look at the binary stream? I saw the one attached by Alex in the other issue but not sure how to get .h264 out of txt/json

**Comment by AlexKordic** [comment]
> O software could detect single-frame segments and fall back to a software transcoder?

I like this idea. Miners could optimize for edge cases by running few software transcoders.

**Comment by MikeIndiaAlpha** [comment]
I looked over the h264 syntax in the file @AlexKordic gave me and found it more or less fine. I mean, apart from duplicated SPS/PPS pair, but that is perfectly legal, even if a bit redundant. Thing is, I saw _a lot_ of problems similar to this one when feeding h.264 decoder on VPU chip of i.mx6 SoC. I think it may be a symptom of a decoder used incorrectly (or alternatively, decoder being really sensitive to the way it is fed with data. I'll be happy to look more into this issue if you want me to.
